### ISSUE TITLE
Wrong types instead of union when fetching API routes having dynamic and static handler

### ISSUE THREAD

**Issue by TheAlexLichter** [nuxt/nuxt#22488]
### Environment

------------------------------
- Operating System: Darwin
- Node Version:     v18.16.1
- Nuxt Version:     3.6.5
- Nitro Version:    2.5.2
- Package Manager:  pnpm@8.5.0
- Builder:          vite
- User Config:      devtools, nitro, runtimeConfig
- Runtime Modules:  -
- Build Modules:    -
------------------------------

### Reproduction

**Important**: Run it locally as StackBlitz limited TS support (for Vue files especially).

https://stackblitz.com/edit/nuxt-22488?file=app.vue,README.md

Manual repro:

1. Create `/server/api/posts/[id].get.ts` and return anything, e.g. `export default defineEventHandler(async () => 42)`
2. Create `/server/api/posts/static.get.ts` and return some thing different, e.g. `export default defineEventHandler(async () => 'static')`
3. Change your app.vue to the following and see the TypeScript "test types" failing, as `data` is not `42 | 'static'` nor `number | string` but `unknown`.

```vue
<script lang="ts" setup>
import type { Ref } from 'vue';

const getId = () => 'random';
const id = getId();

const { data } = await useFetch(`/api/posts/` + id);
//       ^?

export type Equals<X, Y> = (<T>() => T extends X ? 1 : 2) extends <T>() => T extends Y ? 1 : 2
  ? true
  : false;

export type Expect<T extends true> = T;
export type ExpectNot<T extends false> = T;

type _TEST = [
  ExpectNot<Equals<typeof data, Ref<unknown>>>,
  ExpectNot<Equals<typeof data, Ref<any>>>
];
</script>

<template>
  <div>
    {{ data }}
  </div>
</template>
```

### Describe the bug

When having API routes that are targeting the same URL, for example `/api/posts/[id].get.ts` and `/api/posts/static.get.ts` the types of their return values are not inferred correctly when using `$fetch` or `useFetch`. Instead, `unknown` will be returned (or `any` when not using strict mode).

<img width="573" alt="image" src="https://github.com/nuxt/nuxt/assets/640208/2ea68d7f-5404-4880-87e2-12258da49acc">

What *should* be returned is a union of possible types.

### Additional context

[In Nitro, this works as expected](https://stackblitz.com/edit/nuxt-22488-nitro?file=routes%2Findex.ts,nitro.config.ts)
(Also suggested to run locally to check)

<img width="554" alt="image" src="https://github.com/nuxt/nuxt/assets/640208/6c98ab9a-a3c5-44f7-9119-ee916d436250">

**Comment by danielroe** [comment]
I think you need to do `/api/posts/${id}` and not `/api/posts/` + id or TS can't use string literal inference.

**Comment by manniL** [comment]
> I think you need to do `/api/posts/${id}` and not `/api/posts/` + id or TS can't use string literal inference.

That's what I tried initially 😋 
Unfortunately with the same result

<img width="551" alt="image" src="https://github.com/nuxt/nuxt/assets/640208/f3b3dbe1-a95e-4322-be98-c2ef339b56f5">

**Comment by danielroe** [comment]
Ah, okay. (Didn't test locally.)

**Comment by shaun-fitzgibbon** [comment]
I have also tried this and it seems to not infer the types from api routes, I have tried all the examples in the [serialization section](https://nuxt.com/docs/getting-started/data-fetching#serialization) of the Nuxt docs and none of them seem to infer types from the api using useFetch or $fetch

Also I cannot get the superjson example to work without setting the response header to  
`event.node.res.setHeader("Content-Type", "text/plain");`

Using Version: Nuxt 3.7.2 with Nitro 2.6.3

**Comment by genu** [comment]
Is there a temporary workaround for this until a fix comes?

**Comment by genu** [comment]
For anyone reading this, a dirty workaround for this is:

On the client:
```ts
const { data } = await useFetch(`/api/project/:${id}`)
```
notice the `:` before the `${id}` That should properly give you the correct return type.

In the api endpoint, you then have to do:

```ts
export default defineEventHandler(async (event) => {
  const id = getRouterParam(event, 'id')?.split(':')[1]!

  // ...
}
```

Millage may vary without proper validation, but it works.

**Comment by mubaidr** [comment]
> notice the : before the ${id} That should properly give you the correct return type.

Have to try this, but if this works, then maybe `getRouterParam` should remove leading `:` by default.

**Comment by th1m0** [comment]
> > I think you need to do `/api/posts/${id}` and not `/api/posts/` + id or TS can't use string literal inference.
> 
> That's what I tried initially 😋 Unfortunately with the same result
> 
> <img alt="image" width="551" src="https://private-user-images.githubusercontent.com/640208/258381739-f3b3dbe1-a95e-4322-be98-c2ef339b56f5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU4NjE1ODAsIm5iZiI6MTczNTg2MTI4MCwicGF0aCI6Ii82NDAyMDgvMjU4MzgxNzM5LWYzYjNkYmUxLWE5NWUtNDMyMi1iZTk4LWMyZWYzMzliNTZmNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMTAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDEwMlQyMzQxMjBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yY2JjYzQ0YTk5ZWIyMjA1ZmVkOTQ3YjY1MGY2NmE0NTY2YzljZmE0MDk4MmQ0MDU2YWU5Nzg3OTBlN2U0YzRhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.6oV0Ton9mRmB9bV6t4H1ljrMci5_VxspN1MYihAQIxs">

This may have been fixed already? 

![Image](https://github.com/user-attachments/assets/23b00888-b0bb-41a4-8276-66dc6eae2051)

Although using the following code will not work as typescript can't use it as a string literal as Daniel said above.

```ts
const { data } = await useFetch(`/api/posts/` + id)
```

**Comment by phat-dang** [comment]
I got the same problem in this discussion: https://github.com/nuxt/nuxt/discussions/32692

**Comment by geyermichael** [comment]
I came across this issue after upgrading to nuxt v4. 
Fetching data using template literal urls seems to raise type issues, and I thought using it like this was the expected way:

```javascript
await useFetch('/api/foo/:id' , { params: { id } })
```

But params are just an alias for query. @danielroe is this intentional?

**Comment by TheAlexLichter** [comment]
@geyermichael Yes, intentional. This follows the ofetch [`query`/`params`](https://github.com/unjs/ofetch#%EF%B8%8F-adding-query-search-params) option. There we discussions some time ago to unify the naming as params is indeed ambiguous (cc @pi0).

**Comment by dword-design** [comment]
> This is still an issue. My current workaround is:
> 
> await useFetch(`/api/test/${(props.productId as ':id')}`,

Looks like this is the best workaround so far.

**Comment by stigvanbrabant** [comment]
I'm bumping into this atm with nuxt version 4.4.2. Are there plans to fix this sometime soon?
